Refactor Batch TriggeredAbility

Susucre opened the pull request:
This is doing in a very limited scope the changes explained #12095, in order to have Felix Five-Boots (and next the others NUMBER_OF_TRIGGERS triggers) work properly with batch events.

TODO:
- [x] Add in verify tests that Triggered Ability checking batch events implement the BatchTriggeredAbility interface. (That's the intent with the boolean added on EventType)
- [x] Do the same change for others Batch events
- [x] Rebase and rework the two MILLED Batch events
- Outside of scope of current PR, fix the other NUMBER_OF_TRIGGERS using the Felix' code pattern.

Note: I have Felix's PR code here (tweaked with the changes), but just to make sure the refactor actually works (it does). Can clean that part if want to merge separately.

Fix #12193

Susucre commented:
New verify test check: non-BatchTriggeredAbility should not accept the EventType that are flagged as batch ones.

![image](https://github.com/magefree/mage/assets/34709007/068941e1-4e8b-4af8-a629-07ccb8c3b994)

xenohedron commented:
I like the direction here.

Susucre commented:
Alright, all batch events have been reworked.

I plan to add some tests in the next few days, in particular for the triggers I've consolidated into common class.

xenohedron commented on Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java:
```diff
@@ -117,12 +104,52 @@ public boolean checkEventType(GameEvent event, Game game) {
+                            return StaticFilters.FILTER_PERMANENT_CREATURE
+                                    .match(permanent, getControllerId(), this, game);
+                        })
+                        .isPresent()
```
this looks wrong - the isPresent() will return true whether the Boolean is true or false, as long as the permanent is not null, which I don't think is what you want.

Susucre replied:
Oh right, need `FILTER_CONTROLLED_PERMANENT_CREATURE`.

This is not about the Optional::filter logic right?

xenohedron replied:
No, I didn't notice the controller aspect. I haven't tested but I'm pretty sure your Optional usage won't work as intended.

Susucre replied:
Isn't that the expected behavior in Java 8?
```
System.out.println(Optional.of(2).filter(n > 1).isPresent()); // true
System.out.println(Optional.of(0).filter(n > 1).isPresent()); // false
```

Susucre replied:
![image](https://github.com/magefree/mage/assets/34709007/76f59504-6fb0-4e0d-bb7e-8a49d2f01b17)

xenohedron replied:
My apologies then, guess I have to go back and read the docs

JayDi85 replied:
There were multiple times with Optional misunderstand already (in all use cases it was used as alternative to “if not null then execute some code”). I don’t like it too.

xenohedron replied:
There's nothing wrong here, Optional is fine style (and often cleaner). I was reading too quickly and made a faulty assumption.

xenohedron commented:
I wonder if it would be cleaner to have a separate "check event" method that can be a single `filter` call in the stream, rather than nesting all the logic

Susucre commented on Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java:
```diff
@@ -117,12 +104,52 @@ public boolean checkEventType(GameEvent event, Game game) {
+                        .of(e)
+                        .map(ZoneChangeEvent::getTargetId)
+                        .map(game::getCard)
+                        .filter(card -> StaticFilters.FILTER_CARD_CREATURE.match(card, getControllerId(), this, game))
```
same here, need to check it's a card you own that is moving.

Susucre commented:
Added some tests (extended the Test Suite to be allow to choose the players for Sower of Discord), and fixed Kaya & wrong sacrifice effect for Phyrexian Totem & Phyrexian Negator.

Should be good to merge. Can merge the Felix's PR first, it is altered here to fix it. The final step will be done later on to fix the 10-ish NUMBER_OF_TRIGGERS cards just like Felix, to be working properly with batch events.

Susucre commented:
Uhm so DelayedTriggeredAbility probably need the same kind of work around Batch Events.
For instance [[Jace, Cunning Castaway]] +1's is broken since it checks the batch event source and not the individual events.

Susucre commented:
Alright, was only missing two due to some blind spot of the Verify Test: Jace, Cunning Castaway and [[Tamiyo, Field Researcher]].
I did look manually at all cards having their own `DelayedTriggerAbility`, so hopefully did not miss any other. I like those sanity checks to be automated as much as possible, but here there is no easy solution except with another big refactor, so way out of scope.

xenohedron commented on Mage/src/main/java/mage/abilities/BatchTriggeredAbility.java:
```diff
@@ -0,0 +1,29 @@
+     * <p>
+     * Note: if calling this, make sure the event does pass TriggeredAbility::checkEventType
+     */
+    Stream<T> filterBatchEvent(GameEvent event, Game game);
```
I would prefer if we could split the stream logic and the filter logic into separate methods, something like:

```java

boolean checkEvent(T event, Game game);

default Stream<T> filterBatchEvent(BatchEvent<T> event, Game game) {
    return event.getEvents()
                .stream()
                .filter(e -> checkEvent(e, game))
}

```

with the checkTrigger methods starting with

`filterBatchEvent((TypeOfBatchEvent) event, game)`

but I'm not sure, will the generic types work for this sort of usage?

Susucre replied:
Uhm I'll try something like that. May need another method to do the event casting a safe way, as the NUMBER_OF_TRIGGERS will not have direct access to the batch event type.

JayDi85 replied:
Yes, must use overrideable method like `filterEvent` from xenohedron above -- so all implementation will override only simple boolean method and no needs in streams at all. Also I recommends to return List<GameEvent> in find method, not stream<T> -- it's a standard code style for xmage.

xenohedron commented on Mage/src/main/java/mage/abilities/common/CombatDamageDealtToYouTriggeredAbility.java:
```diff
@@ -71,20 +76,25 @@ public boolean checkEventType(GameEvent event, Game game) {
+                .getEvents()
+                .stream()
+                .filter(DamagedPlayerEvent::isCombatDamage)
+                .filter(e -> e.getPlayerId() == getControllerId());
```
compare uuid with .equals

Susucre replied:
I usually get those right those days. Good catch.

xenohedron commented on Mage/src/main/java/mage/abilities/common/OneOrMoreDealDamageTriggeredAbility.java:
```diff
@@ -59,12 +61,12 @@ public boolean checkEventType(GameEvent event, Game game) {
-    public boolean checkTrigger(GameEvent event, Game game) {
+    public Stream<DamagedPlayerEvent> filterBatchEvent(GameEvent event, Game game) {
         DamagedBatchForOnePlayerEvent dEvent = (DamagedBatchForOnePlayerEvent) event;
         if (onlyCombat && !dEvent.isCombatDamage()) {
```
wouldn't it make more sense for these sorts of checks (on info about the batch event rather than the sub events) to be in `checkTrigger` before calling this method for the stream?

Susucre replied:
Yeah I'll do a cleanup pass for the combat damage, and the target (resp. source) when the Batch event is one that has a shared target (resp. source)

JayDi85 commented on Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java:
```diff
@@ -2426,15 +2447,31 @@ public boolean chooseTarget(Outcome outcome, Target target, Ability source, Game
                         continue;
                     }
-                    checkTargetDefinitionMarksSupport(target, targetDefinition, "=");
-                    String playerName = targetDefinition.substring(targetDefinition.indexOf("targetPlayer=") + 13);
```
1. Old code used in 3 places, but changed in one. Looks like it miss for `handlePlayerTarget`:
![shot_240521_170617](https://github.com/magefree/mage/assets/8344157/b53774f7-e8df-4dc0-ab3c-a4ed3b3ec610)

2. I recommends to use real targets lifecycle in tests by `addTarget` commands, not multi-targets by `^` (the last one added it directly and can be potentially broken in dynamic targets):
![shot_240521_171301](https://github.com/magefree/mage/assets/8344157/e325e4a3-2511-42a2-b5b1-b72006d2404a)

JayDi85 commented on Mage.Tests/src/test/java/org/mage/test/serverside/base/impl/CardTestPlayerAPIImpl.java:
```diff
@@ -2148,6 +2154,21 @@ public void addTarget(TestPlayer player, TestPlayer targetPlayer, int timesToCho
+        String target = "targetPlayer=";
+        for (int i = 0; i < targetPlayers.length; ++i) {
+            if (i > 0) {
+                target += "^";
```
That method must be removed can cause bugs (see above about dynamic targets). Use addTarget per player.

JayDi85 replied:
@Susucre are you planning to fix that?

JayDi85 replied:
Related issue #11666

JayDi85 commented on Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java:
```diff
@@ -1719,6 +1719,10 @@ private void compareClassRecursive(Object obj1, Object obj2, Card originalCard,
                 }
 
+                if (obj1 instanceof Ability) {
+                    checkAbility(originalCard, (Ability) obj1, msg);
```
`compareClassRecursive` is a copy code checker. Use `checkMissingAbilities` instead and put special check in it. Example:

![shot_240521_172054](https://github.com/magefree/mage/assets/8344157/fc244ff2-b964-4538-aabd-6721d69084af)

Susucre replied:
for that test to do anything, it needs to attempt to deep recurse down the abilities to find the triggers. `checkMissingAbilities` is doing text at high level.

JayDi85 commented on Mage/src/main/java/mage/abilities/common/BecomesTappedOneOrMoreTriggeredAbility.java:
```diff
@@ -37,12 +42,21 @@ public boolean checkEventType(GameEvent event, Game game) {
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        return filterBatchEvent(event, game).findAny().isPresent();
```
See above about List. Example:
* `BatchTriggeredAbility` extends `TriggeredAbilityImpl`;
  * it's has default method `filterEvent` with `return true`;
  * it's has final method like `List findFilteredEvents(Event, Game)` that calls implementations of `filterEvent`;
  * it's has default implementation of `checkTrigger` that `return !findFilteredEvents(event, game).isEmpty`;
* All implementations of `BatchTriggeredAbility`:
  *  overrides `filterEvent` method to filter only needed;
  *  override `checkTrigger` method to setup dynamic targets or other additional code.

That's all will work same way as other triggered abilities, but with additional middleware (filterEvent);

P.S. I don't like generic usages all around. If it possible to use standard classes then it will be better (extends basic ability, extends basic event, etc). If it (generic) can simplify some code then can be keep.

Susucre replied:
Well I won't do all that additional work that doesn't accomplish much.
So either someone else take the PR from here, or we can close it and leave the bugs.

xenohedron replied:
I guess I'll work on that then.

JayDi85 reviewed (commented):
Checked game engine changes only (without abilities logic). I prefer to use a shared  BatchTriggeredAbility class instead interface. See related comments with example.

Susucre commented on Mage.Tests/src/test/java/org/mage/test/cards/single/c18/SowerOfDiscordTest.java:
```diff
@@ -0,0 +1,141 @@
+        addCard(Zone.BATTLEFIELD, playerA, "Badlands", 7);
+
+        castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, sower, true);
+        addTarget(playerA, playerB, playerD);
```
@JayDi85 tell me how to do those tests without the extra support for `^` on targetting player.

This is the target for a trigger, and it needs to target multiple players.
